docs: add first guide for schema governance series (#2267)
graphql/graphql.github.io: src/pages/learn/schema-review.mdx

## Schedule periodic reviews

Regular schema reviews create opportunities to address accumulated issues and align on standards.

### Schedule periodic schema health reviews

Beyond reviewing individual changes, examine your entire schema quarterly. Look for:

- **Unused types and fields**: Candidates for deprecation
- **Inconsistent patterns**: Areas needing standardization
- **Missing documentation**: Types and fields that need descriptions
- **Performance bottlenecks**: Overly complex resolvers
- **Consolidation opportunities**: Similar types that could merge

Create a schema health scorecard that tracks these metrics over time.

```javascript
export async function generateSchemaHealthReport(schema, usageData) {
  const types = Object.keys(schema.getTypeMap())
    .filter(name => !name.startsWith('__'));
  
  const fields = types.flatMap(typeName => {
    const type = schema.getType(typeName);
    if ('getFields' in type) {

>       return Object.keys(type.getFields());

Reviewer: I'm not certain how important this is but, this part produces an array like `["id", "name", "price"]`, just bare field names with no type prefix. Then later, it looks up `usageData.get("name")` instead of `usageData.get("Product.name")`. So if two types share a field name, they'd collide in the lookup.

```suggestion
      return Object.keys(type.getFields()).map(f => `${typeName}.${f}`);
```

      "requiredApprovers": ["@api-leadership"]
    },
    "shared-type": {
      "requiredApprovers": ["@api-platform-team"]
    }
  }
}
```

This example shows configuration for automated approval enforcement. The specific rules matter less than making them explicit and consistently applied.

To reduce approval burden, make most changes non-breaking. Add new fields instead of modifying existing ones, use deprecation instead of removal, and introduce new types rather than changing established ones. Reserve the heavyweight approval process for changes that genuinely require broad consensus.

## Manage breaking changes

Breaking changes sometimes become necessary despite best efforts to avoid them. When you must make breaking changes, follow a deliberate process that minimizes client impact.

### Identify what constitutes a breaking change

Breaking changes include:

- **Removing fields or types**: Clients requesting removed fields will fail
- **Removing or renaming arguments**: Queries using old argument names break
- **Changing field types**: Type mismatches cause client errors
- **Adding required arguments**: Existing queries missing new required arguments fail

> - **Removing enum values**: Queries requesting removed values break

Reviewer 2: I'd like to get more precision there. This is quite a big topic. Suggestion to:

1. add a definition of a breaking change
2. use consistent wording (existing operations, existing clients)

```suggestion
A breaking change is any change that breaks the server/client contract. This happens when previously valid operations become invalid or when the shape of the returned data changes.

Breaking changes include:

- **Removing fields or types**: Existing operations requesting removed fields become invalid
- **Removing or renaming arguments**: Existing operations using old argument names become invalid
- **Adding required arguments**: Existing operations missing new required arguments become invalid
- **Removing enum values**: Existing operations using removed values in input positions become invalid
- **Changing a field type**: Existing clients may crash on the unexpected type
```

> - **Making nullable fields non-nullable**: Clients expecting null values may break

Reviewer 2: Other way around:

```suggestion
- **Making non-nullable fields nullable**: Existing clients may crash on unexpected nulls
```

> 

Reviewer 2: Add some wording that a breaking change doesn't always mean a broken application:

```suggestion

Note that some of these changes, such as a type rename or a conversion from an object to an interface, may be compatible with existing clients at runtime. But it is impossible to tell for sure and are therefore considered breaking.
```

> 
> Dangerous changes that may break some clients include:
> 
> - **Adding enum values**: If clients exhaustively match enum values
> - **Adding fields to input types**: If clients use spread operators to construct input objects

Reviewer 2: I would very much like that we change the narrative from "this is dangerous by default" to "this should be safe unless your frontend doesn't support it".

```suggestion
Some changes may break applications that did not plan for schema evolution:

- **Adding enum values**: If application code exhaustively match enum values
- **Adding new types to an union or interface**: If application code exhaustively match the type name
- **Adding fields to input types**: If application code use spread operators to construct input objects

Such applications make schema evolution harder and writing robust applications is encouraged. 

If in doubt, use caution when doing those changes.
```

I wrote a "Robust Applications" best practice PR [there](https://github.com/graphql/graphql.github.io/pull/2434) that we could then link from here.

> Codegen tools can break on schema changes that are technically backwards compatible. Adding a new field to a type may cause TypeScript codegen to fail if existing queries don't request that field, depending on how the codegen tool handles schema evolution.

Reviewer 2: I'm not sure I understand this one. Did you have an example?

> Codegen tools can break on schema changes that are technically backwards compatible. Adding a new field to a type may cause TypeScript codegen to fail if existing queries don't request that field, depending on how the codegen tool handles schema evolution.

Reviewer 2: Link to https://graphql.org/learn/robust-applications/

```suggestion
Dangerous changes that may break some clients include:

- **Adding enum values**: If clients exhaustively match enum values
- **Adding new types to a union/interface**: If clients exhaustively match the cases
- **Adding fields to input types**: If clients use spread operators to construct input objects

Modern clients should handle those cases (see [./robust-applications]) but some older clients may break and caution is encouraged for those changes. 
```
